### PR TITLE
fix(kubernetes_logs source): Fix events being empty when log namespacing is enabled

### PR DESCRIPTION
**Pull request by fuchsnj**
closes https://github.com/vectordotdev/vector/issues/18018

The `kubernetes_logs` source internally used the `reduce` transform to merge partial logs. There is an [issue](https://github.com/vectordotdev/vector/issues/18219) in the `reduce` transform where it doesn't work with events where the root of the event is not an object (which only happens with log namespacing enabled).

After a short discussion with the team, we decided to stop using `reduce` internally and instead use some custom code here.

A new function, `map_with_expiration` (I'm open to better name suggestions here) was created to encapsulate some of the more complicated async stream operations of the `reduce` transform. This was used in the new `kubernetes_logs` source, and the `reduce` transform was also re-written to use it.

### CODE REVIEW

File: src/sources/kubernetes_logs/path_helpers.rs [hunk 1]
```diff
@@ -37,7 +37,7 @@ pub(super) fn parse_log_file_path(path: &str) -> Option<LogFileInfo<'_>> {
     let mut components = path.rsplit('/');
 
     let _log_file_name = components.next()?;
-    let container_name = components.next()?;
+    // let container_name = components.next()?;
```
[the remark this answers is not in the document]
Comment (fuchsnj) [reply src/sources/kubernetes_logs/path_helpers.rs]: fixed

File: lib/vector-core/src/stream/expiration_map.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                                map_fn(&mut state, event, &mut emitter);
+                          false
+                        }
+                      }
+                    }
+                };
+                yield futures::stream::iter(emitter.values.into_iter());
+                if done { break }
+              }
+
```
Comment (spencergilbert) [inline lib/vector-core/src/stream/expiration_map.rs:65]: Complaints about formatting... would be nice if the formatter worked inside macros.

**Review by spencergilbert — APPROVED** [review]
LGTM, primarily focused on the tests in the k8s source.

**Review by pront — APPROVED** [review]
LGTM! 

Note: rebase on latest master for the `cargo-deny` fix.

**Pull request merged**